src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -0,0 +1,337 @@
+                                //an auto-referenced package was printed
+                                ProjectPackagesPrintUtility.PrintPackages(packages, projectName, listPackageArgs.IncludeTransitive, listPackageArgs.IncludeOutdated, out var autoRefFoundWithinProject);
+                                autoReferenceFound = autoReferenceFound || autoRefFoundWithinProject;
+                            }
+                        }
+
+                    }
+                    else
+                    {
+                        Console.WriteLine(string.Format(Strings.ListPkg_ErrorReadingAssetsFile, assetsPath));
+                    }
+                    //Unload project
+                    ProjectCollection.GlobalProjectCollection.UnloadProject(project);
+                }
+            }
+
+            //If any auto-references were found, a line is printed
+            //explaining what (A) means
+            if (autoReferenceFound)
+            {
+                Console.WriteLine(Strings.ListPkg_AutoReferenceDescription);
+            }
+        }
+
+        /// <summary>
+        /// Fetches the latest versions for all of the packages that are
+        /// to be listed
+        /// </summary>
+        /// <param name="packages">The packages found in a project</param>
+        /// <param name="listPackageArgs">List args for the token and source provider</param>
+        /// <returns>A data structure like packages, but includes the latest versions</returns>
+        private async Task AddLatestVersionsAsync(
+            IEnumerable<FrameworkPackages> packages, ListPackageArgs listPackageArgs)
+        {
+            var providers = Repository.Provider.GetCoreV3();
+
+            //Handling concurrency and throttling variables
+            var maxTasks = Environment.ProcessorCount;
+            var contactSourcesRunningTasks = new List<Task>();
+            var latestVersionsRequests = new List<Task>();
+
+            //Unique Dictionary for packages and list of latest versions to handle different sources
+            var packagesVersionsDict = new Dictionary<InstalledPackageReference, IList<NuGetVersion>>(new InstalledPackageReferenceIdComparer());
+            AddPackagesToDict(packages, packagesVersionsDict);
+
+            //Prepare requests for each of the packages
+            foreach (var package in packagesVersionsDict)
+            {
+                PrepareLatestVersionsRequests(package.Key, listPackageArgs, providers, latestVersionsRequests, packagesVersionsDict);
+            }
+
+            //Make the calls to the sources
+            foreach (var latestVersionTask in latestVersionsRequests)
+            {
+                contactSourcesRunningTasks.Add(latestVersionTask);
+                //Throttle if needed
+                if (maxTasks <= contactSourcesRunningTasks.Count())
+                {
+                    var finishedTask = await Task.WhenAny(contactSourcesRunningTasks);
```

Comment:
where have you started the task `latestVersionTask`?

Reply:
`WhenAny` and `WhenAll` start all of the tasks in the list. I should start the tasks once they are added instead of waiting for them to reach the limit and launching them